hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/checksum/TestContainerChecksumTreeManager.java

```diff
@@ -0,0 +1,169 @@
+    when(container.getMetadataPath()).thenReturn(testDir.getAbsolutePath());
+    checksumFile = new File(testDir, CONTAINER_ID + ".tree");
+    checksumManager = new ContainerChecksumTreeManager(new OzoneConfiguration());
+  }
+
+  @Test
+  public void testWriteEmptyTreeToFile() throws Exception {
+    checksumManager.writeContainerDataTree(container, new ContainerMerkleTree());
+    ContainerProtos.ContainerChecksumInfo checksumInfo = readFile();
+
+    assertEquals(CONTAINER_ID, checksumInfo.getContainerID());
+    assertTrue(checksumInfo.getDeletedBlocksList().isEmpty());
+    ContainerProtos.ContainerMerkleTree treeProto = checksumInfo.getContainerMerkleTree();
+    assertEquals(0, treeProto.getDataChecksum());
+    assertTrue(treeProto.getBlockMerkleTreeList().isEmpty());
+  }
+
+  @Test
+  public void testWriteEmptyBlockListToFile() throws Exception {
+    checksumManager.markBlocksAsDeleted(container, new TreeSet<>());
+    ContainerProtos.ContainerChecksumInfo checksumInfo = readFile();
+
+    assertEquals(CONTAINER_ID, checksumInfo.getContainerID());
+    assertTrue(checksumInfo.getDeletedBlocksList().isEmpty());
+    ContainerProtos.ContainerMerkleTree treeProto = checksumInfo.getContainerMerkleTree();
+    assertEquals(0, treeProto.getDataChecksum());
+    assertTrue(treeProto.getBlockMerkleTreeList().isEmpty());
+  }
+
+  @Test
+  public void testWriteOnlyTreeToFile() throws Exception {
+    ContainerMerkleTree tree = buildTestTree();
+    checksumManager.writeContainerDataTree(container, tree);
+
+    ContainerProtos.ContainerChecksumInfo checksumInfo = readFile();
+
+    assertEquals(CONTAINER_ID, checksumInfo.getContainerID());
+    assertTrue(checksumInfo.getDeletedBlocksList().isEmpty());
+    // TestContainerMerkleTree verifies that going from ContainerMerkleTree to its proto is consistent.
+    // Therefore, we can use the proto version of our expected tree to check what was written to the file.
+    assertTreesSortedAndMatch(tree.toProto(), checksumInfo.getContainerMerkleTree());
+  }
+
+  @Test
+  public void testWriteOnlyDeletedBlocksToFile() throws Exception {
+    List<Long> expectedBlocksToDelete = Arrays.asList(1L, 2L, 3L);
+    checksumManager.markBlocksAsDeleted(container, new TreeSet<>(expectedBlocksToDelete));
+
+    ContainerProtos.ContainerChecksumInfo checksumInfo = readFile();
+
+    assertEquals(CONTAINER_ID, checksumInfo.getContainerID());
+    assertEquals(expectedBlocksToDelete, checksumInfo.getDeletedBlocksList());
+    ContainerProtos.ContainerMerkleTree treeProto = checksumInfo.getContainerMerkleTree();
+    assertEquals(0, treeProto.getDataChecksum());
+    assertTrue(treeProto.getBlockMerkleTreeList().isEmpty());
+  }
+
+  @Test
+  public void testDeletedBlocksPreservedOnTreeWrite() throws Exception {
+    List<Long> expectedBlocksToDelete = Arrays.asList(1L, 2L, 3L);
+    checksumManager.markBlocksAsDeleted(container, new TreeSet<>(expectedBlocksToDelete));
```

Comment:
I missed this in the previous PR, we need a test to validate duplicate entries are stored only once in the tree. The interface is purely add a Collection to the deleted list but the API guarantee needs to be tested. Maybe change the API to Set (it is ok to leave as a collection as the API should manage this).

Reply:
Yes I will add this.

Reply:
I added more tests to `TestContainerChecksumTreeManager` for this. What happened is that originally the method required `SortedSet` as input so the extra tests weren't necessary. I made it `Collection` so that the block deleting service could pass anything in and not worry about conversion. The manager would do that in this method. This loosened the contract but I didn't add tests that the internals were still handling it.